python/ray/tune/suggest/search.py

```diff
@@ -0,0 +1,184 @@
+                the RUNNING state and errors.
+            early_terminated (bool): Defaults to False. True if the trial
+                is stopped while in PAUSED or PENDING state.
+        """
+        pass
+
+    def is_finished(self):
+        """Returns True if no trials left."""
+        raise NotImplementedError
+
+
+class ExistingVariants(SearchAlgorithm):
+    """Uses Tune's variant generation for resolving variables.
+
+    Custom search algorithms can extend this class easily by overriding the
+    `try_suggest` method, which will override conflicting fields from
+    the initially generated parameters.
+
+    See `ray.tune.suggest.variant_generator`.
+
+    To track suggestions and their corresponding evaluations, the method
+    `try_suggest` will need to generate a trial_id. This trial_id will
+    be used in subsequent notifications.
+
+    Attributes:
+        PASS (str): Status string for `try_suggest` if ExistingVariants
+            currently cannot be queried for parameters (i.e. due to
+            constrained concurrency).
+
+    Example:
+        >>> suggester = ExistingVariants()
+        >>> new_parameters, trial_id = suggester.try_suggest()
+        >>> suggester.on_trial_complete(trial_id, result)
+        >>> better_parameters, trial_id2 = suggester.try_suggest()
+    """
+
+    PASS = "PASS"
+
+    def __init__(self, experiments=None):
+        """Constructs a generator given experiment specifications.
+
+        Arguments:
+            experiments (Experiment | list | dict): Experiments to run.
+        """
+        exp_list = experiments
+        if isinstance(experiments, Experiment):
+            exp_list = [experiments]
+        elif type(experiments) is dict:
+            exp_list = [
+                Experiment.from_json(name, spec)
+                for name, spec in experiments.items()
+            ]
+        if (type(exp_list) is list
+                and all(isinstance(exp, Experiment) for exp in exp_list)):
+            if len(exp_list) > 1:
+                print("Warning: All experiments will be"
+                      " using the same Search Algorithm.")
+        else:
+            raise TuneError("Invalid argument: {}".format(experiments))
+
+        self._generator = chain.from_iterable([
+            generate_trials(experiment.spec, experiment.name, self)
+            for experiment in exp_list
+        ])
+
+        self._finished = False
+
+    def next_trial(self):
+        try:
+            trial = next(self._generator)
+        except StopIteration:
+            trial = None
+            self._finished = True
+        return trial
+
+    def try_suggest(self):
+        """Queries the algorithm to retrieve the next set of parameters.
+
+        Returns:
+            dict|PASS: Partial configuration for a trial, if possible. This
+                configuration will override conflicting fields from
+                the initially generated parameters. Else, returns
```

Comment:
I don't get why it needs to override fields, why can't suggest() just call super.suggest() and modify it as needed?

I imagine there are three variants of this:
1. A subclass always calls super.suggest(), and then modifies it as needed.
2. A subclass never calls super.suggest but comes up with its own configs.
3. A subclass initially uses super.suggest but after the first wave of trials comes up with its own configs.

Reply:
What is `super.suggest()` supposed to do here?

Reply:
It should return the config from variant generator, that's all that we need right?

It seems to me that producing the actual Trial object is always the same, so we should make that a private method that calls suggest().

Reply:
I see; one thing that has been very useful in the past was specifying `trial_resources:gpus` and `trial_resources:cpus` as a variable parameter, depending on the resolved config (i.e., like `num_workers` for rllib). 

Is the proposal to drop this functionality and only run `generate_variants(unresolved_spec["config"])` rather than `generate_variants(unresolved_spec)`?

Reply:
I see, so fundamentally we need to return full Trials since sometimes you need to tweak the trial runtime config.

In that case I would just have suggest() always return a Trial object, and add a helper function `self.trials_for_config(config)` that produces the trial(s). That way the GPU stuff can still work, and you can also manually construct a trial if needed.